boot/bootutil/src/encrypted_psa.c

```diff
@@ -0,0 +1,465 @@
+        BOOT_LOG_WRN("Key derivation cleanup failed %d", psa_ret);
+    }
+    if (psa_ret != PSA_SUCCESS) {
+        BOOT_LOG_ERR("Key derivation failed %d", psa_ret);
+        return -1;
+    }
+
+    /* The derived key consists of BOOTUTIL_CRYPTO_AES_CTR_KEY_SIZE bytes
+     * followed by BOOTUTIL_CRYPTO_SHA256_DIGEST_SIZE bytes. Both parts will
+     * be imported at the point where needed and discarded immediately after.
+     */
+    psa_set_key_type(&kattr, PSA_KEY_TYPE_HMAC);
+    psa_set_key_usage_flags(&kattr, PSA_KEY_USAGE_VERIFY_MESSAGE);
+    psa_set_key_algorithm(&kattr, PSA_ALG_HMAC(PSA_ALG_SHA_256));
+
+    /* Import the MAC tag key part of derived key, that is the part that starts
+     * after BOOTUTIL_CRYPTO_AES_CTR_KEY_SIZE and has length of
+     * BOOTUTIL_CRYPTO_SHA256_DIGEST_SIZE bytes.
+     */
+    psa_ret = psa_import_key(&kattr,
+                             &derived_key[BOOTUTIL_CRYPTO_AES_CTR_KEY_SIZE],
+                             BOOTUTIL_CRYPTO_SHA256_DIGEST_SIZE, &kid);
+    psa_reset_key_attributes(&kattr);
+    if (psa_ret != PSA_SUCCESS) {
+        memset(derived_key, 0, sizeof(derived_key));
+        BOOT_LOG_ERR("MAC key import failed %d", psa_ret);
+        return -1;
+    }
+
+    /* Verify the MAC tag of the random encryption key */
+    psa_ret = psa_mac_verify(kid, PSA_ALG_HMAC(PSA_ALG_SHA_256),
+                             &buf[EC_CIPHERKEY_INDEX], BOOT_ENC_KEY_SIZE,
+                             &buf[EC_TAG_INDEX],
+                             BOOTUTIL_CRYPTO_SHA256_DIGEST_SIZE);
+    psa_cleanup_ret = psa_destroy_key(kid);
+    if (psa_cleanup_ret != PSA_SUCCESS) {
+        BOOT_LOG_WRN("MAC key destruction failed %d", psa_cleanup_ret);
+    }
+    if (psa_ret != PSA_SUCCESS) {
+        memset(derived_key, 0, sizeof(derived_key));
+        BOOT_LOG_ERR("MAC verification failed %d", psa_ret);
+        return -1;
+    }
+
+    /* The derived key is used in AES decryption of random key */
+    psa_set_key_type(&kattr, PSA_KEY_TYPE_AES);
+    psa_set_key_usage_flags(&kattr, PSA_KEY_USAGE_DECRYPT);
+    psa_set_key_algorithm(&kattr, PSA_ALG_CTR);
+
+    /* Import the AES partition of derived key, the first 16 bytes */
+    psa_ret = psa_import_key(&kattr, &derived_key[0],
+                             BOOTUTIL_CRYPTO_AES_CTR_KEY_SIZE, &kid);
+    memset(derived_key, 0, sizeof(derived_key));
+    if (psa_ret != PSA_SUCCESS) {
+        BOOT_LOG_ERR("AES key import failed %d", psa_ret);
+        return -1;
+    }
+
+    /* Decrypt the random AES encryption key with AES and the key obtained
+     * at derivation. */
+    memset(&iv_and_key[0], 0, PSA_CIPHER_IV_LENGTH(PSA_KEY_TYPE_AES, PSA_ALG_CTR));
+    memcpy(&iv_and_key[PSA_CIPHER_IV_LENGTH(PSA_KEY_TYPE_AES, PSA_ALG_CTR)],
+           &buf[EC_CIPHERKEY_INDEX],
+	   sizeof(iv_and_key) - PSA_CIPHER_IV_LENGTH(PSA_KEY_TYPE_AES, PSA_ALG_CTR));
+
+    /* QQQ: are we sure that enckey has the size proper to fit the key ? */
```

Comment:
should these comments be addressed/looked at?

Reply:
Removed. The comments are note for improvement, where MCUboot has very poor "management" of various identifiers/constants that are hard to track.